### ISSUE TITLE
Is ImageUploader.js supposed to work with PHP?

### ISSUE THREAD

**Issue by that-ben** [rossturner/HTML5-ImageUploader#5]
Hi! I've given a go to this project in hope of replacing the only little bit of Flash left on a project I'm doing at the moment.  I grabbed a copy of the demo and changed the uploadUrl variable.  I see the POST headers sent via Firebug and they contain DATA.  It looks like the image is uploaded but on the PHP side of it, $_POST array is completely empty.  I also listed the contents of $_FILES and $_GET arrays just to make sure and they're completely empty too.

Is there something I'm missing?

**Comment by rossturner** [comment]
Hi Ben,

It's a pretty plain HTTP request sending the image data as a base64-encoded string as the body of the request. I believe `$_POST` in PHP is designed to be used for a POST request from submitting a form (from http://php.net/manual/en/reserved.variables.post.php). In this case it's a POST with essentially a big string as the body of the POST request, which needs interpreting by the server in some way.

I'm afraid I've no real experience with PHP to point you in the right direction there, though I'm sure it's possible to either turn the submission into a form by sticking the image data into a hidden input, or (preferably) doing something server-side to handle a non-form POST request. Perhaps something like http://stackoverflow.com/questions/8945879/how-to-get-body-of-a-post-in-php ?

**Comment by that-ben** [comment]
First, thanks for replying.  It was very quick too! OK, I'll have a look at your JS file and I'll post a solution here for the future users of ImageUploader.js because that would be VERY useful if your script was compatible with PHP.  At least 80% of websites use that technology to pre-process web pages.  PHP is everywhere.  This needs to work with it, not only with Java :)

**Comment by rossturner** [comment]
Yeah Java is a bit of a weird choice for this one - put it down to a lack of experience 5 years ago! I'd probably write this as a pure JS thing backed by Node if I was to redo it today. 

Really the repo was a proof of concept for the time period as my technical lead didn't think a pure JS image upload and resize was possible at that point in time ;)

**Comment by that-ben** [comment]
Well, you beat me to it, actually.  It's true all we have to do to make it work with PHP is to read the input stream in our PHP script instead of reading the $_POST array.  Here's the result of reading various input variables:

_POST: array(0) { }
_FILES: array(0) { }
_GET: array(0) { }
file_get_contents('php://input');: string(889340) "/9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAAEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQH/2wBDAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQH/wAARCAKHAw4DAREAAhEBAxEB/8QAHwAAAQUBAQEBAQ
[...]

So basically, after copying the input stream into a variable with: file_get_contents('php://input'); we only need to Base64 decode it and then we obtain the JPEG image data as expected.  In a nutshell, here's the full PHP code to make it work:

// Read input stream
$raw_64 = file_get_contents('php://input');

// Base64 decode the input stream
$img = base64_decode($raw_64);

// Save the JPEG image on the server disk
file_put_contents('__test.jpg', $img);

**Comment by rossturner** [comment]
Glad you sorted it :) I don't mind leaving this open in case someone can stumble across it to help them integrate with PHP. I'd like to go back to this project and clean it up and improve it quite a bit over a 5 year old version, I just can't find the time when I'm trying to get some game development done in any free time I come up with!

**Comment by that-ben** [comment]
BTW, ImageUploader.js is very nice.  At last I'll be free of Flash :) HTML5 rocks ;-)

**Comment by rossturner** [comment]
Thanks, that was exactly why I wanted it in the past too, except even worse in that instance - we were tied up with Silverlight!

**Comment by rvalitov** [comment]
Issue can be closed, a full example including PHP code is now available in README file #6.

**Comment by rossturner** [comment]
Thanks a lot for the great PR :)
